p2p/kademlia/network.go

```diff
@@ -0,0 +1,286 @@
+
+	// add the sender to local hash table
+	s.dht.addNode(message.Sender)
+
+	// format the key
+	key := s.dht.hashKey(request.Data)
+	// replication time for key
+	replication := time.Now().Add(defaultReplicateTime)
+	// store the data to local storage
+	if err := s.dht.store.Store(ctx, key, request.Data, replication); err != nil {
+		return nil, errors.Errorf("store the data: %w", err)
+	}
+
+	// new a response message
+	response := s.dht.newMessage(StoreData, message.Sender, nil)
+	// send the response to client
+	encoded, err := encode(response)
+	if err != nil {
+		return nil, errors.Errorf("encode response for store data: %w", err)
+	}
+
+	return encoded, nil
+}
+
+func (s *Network) handlePing(_ context.Context, message *Message) ([]byte, error) {
+	// new a response message
+	response := s.dht.newMessage(Ping, message.Sender, nil)
+	// send the response to client
+	encoded, err := encode(response)
+	if err != nil {
+		return nil, errors.Errorf("encode response for ping: %w", err)
+	}
+
+	return encoded, nil
+}
+
+// handle the connection request
+func (s *Network) handleConn(ctx context.Context, conn net.Conn) {
+	defer conn.Close()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		default:
+		}
+
+		// read the request from connection
+		request, err := decode(conn)
+		if err != nil {
+			if err == io.EOF {
+				return
+			}
+			log.WithContext(ctx).Errorf("read and decode: %v", err)
+			return
+		}
+
+		var response []byte
+		switch request.MessageType {
+		case FindNode:
+			encoded, err := s.handleFindNode(ctx, request)
+			if err != nil {
+				log.WithContext(ctx).Errorf("handle find node request: %v", err)
+				continue
+			}
+			response = encoded
+		case FindValue:
+			// handle the request for finding value
+			encoded, err := s.handleFindValue(ctx, request)
+			if err != nil {
+				log.WithContext(ctx).Errorf("handle find value request: %v", err)
+				continue
+			}
+			response = encoded
+		case Ping:
+			encoded, err := s.handlePing(ctx, request)
+			if err != nil {
+				log.WithContext(ctx).Errorf("handle ping request: %v", err)
+				continue
+			}
+			response = encoded
+		case StoreData:
+			// handle the request for storing data
+			encoded, err := s.handleStoreData(ctx, request)
+			if err != nil {
+				log.WithContext(ctx).Errorf("handle store data request: %v", err)
+				continue
+			}
+			response = encoded
+		default:
+			log.WithContext(ctx).Errorf("impossible: invalid message type: %v", request.MessageType)
+		}
+
+		// write the response
+		if _, err := conn.Write(response); err != nil {
+			log.WithContext(ctx).Errorf("conn write: %v", err)
+		}
+	}
+}
+
+// serve the incomming connection
+func (s *Network) serve(ctx context.Context) {
+	for {
+		// accept the incomming connections
+		conn, err := s.socket.Accept()
+		if err != nil {
+			if strings.Contains(err.Error(), "closed") {
+				return
+			}
+
+			log.WithContext(ctx).Errorf("socket accept: %v", err)
```

Comment:
in case of socket error (maybe due to network error), if we stop here then we will lost connection to other nodes.
Do you think should try to restart network again?

Reply:
Agree, i add the logic according to the http server's accept